File: frontend/segmentation_fault/api/posts.ts
import { get, httpDelete, post as httpPost, put } from ".";
import { externalisePostUpdate, internaliseComment, internalisePost } from "./conversions";
import { Comment, CommentExternal, Post, PostCreate, PostExternal, PostUpdate } from "./schemas";

const ROUTE_PREFIX = "/posts";

export async function getPosts(): Promise<Post[]> {
    const [status, posts] = await get<PostExternal[]>(`${ROUTE_PREFIX}/`);

    if (status === 200) {
        return await Promise.all(posts.map(internalisePost));
    } else {
        throw new Error("Unknown error");
    }
}

export async function createPost(post: PostCreate): Promise<Post> {
    const [status, createdPost] = await httpPost<PostExternal>(
        `${ROUTE_PREFIX}/`,
        JSON.stringify(post)
    );

    if (status === 201) {
        return internalisePost(createdPost);
    } else {
        throw new Error("Unknown error");
    }
}

export async function getPost(postId: number): Promise<Post> {
    const [status, post] = await get<PostExternal>(`${ROUTE_PREFIX}/${postId}`);

    if (status === 200) {
        return internalisePost(post);
    } else if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else {
        throw new Error("Unknown error");
    }
}

export async function updatePost(postId: number, post: PostUpdate): Promise<Post> {
    const [status, updatedPost] = await put<PostExternal>(
        `${ROUTE_PREFIX}/${postId}`,
        JSON.stringify(externalisePostUpdate(post))
    );

    if (status === 200) {
        return internalisePost(updatedPost);
    } else if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else {
        throw new Error("Unknown error");
    }
}

export async function deletePost(postId: number): Promise<void> {
    const [status, _] = await httpDelete<PostExternal>(`${ROUTE_PREFIX}/${postId}`);

    if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else if (status !== 200) {
        throw new Error("Unknown error");
    }
}

export async function getPostComments(postId: number): Promise<Comment[]> {
    const [status, children] = await get<CommentExternal[]>(`${ROUTE_PREFIX}/${postId}/comments`);

    if (status === 200) {
        return await Promise.all(children.map(internaliseComment));
    } else if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else {
        throw new Error("Unknown error");
    }
}

export async function getPostVotes(postId: number): Promise<number> {
    const [status, votes] = await get<number>(`${ROUTE_PREFIX}/${postId}/votes`);

    if (status === 200) {
        return votes;
    } else if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else {
        throw new Error("Unknown error");
    }
}

export async function getPostVote(postId: number): Promise<boolean | null> {
    const [status, vote] = await get<string>(`${ROUTE_PREFIX}/${postId}/vote`);

    if (status === 200) {
        return JSON.parse(vote);
    } else if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else {
        throw new Error("Unknown error");
    }
}

export async function createPostVote(postId: number, up: boolean | null): Promise<void> {
    const [status, _] = await httpPost(`${ROUTE_PREFIX}/${postId}/vote?type=${up}`);

    if (status === 404) {
        throw new Error(`Post '${postId}' not found`);
    } else if (status !== 201) {
        throw new Error("Unknown error");
    }
}

File: backend/segmentation_fault/routes/posts.py
import controllers.posts as controller
from controllers.auth import get_current_user
from fastapi import APIRouter, Depends, status
from schemas.comments import Comment
from schemas.posts import Post, PostCreate, PostUpdate
from schemas.users import User
from schemas.votes import VoteType

router = APIRouter(dependencies=[Depends(get_current_user)])


@router.get("/", response_model=list[Post])
async def get_posts():
    return await controller.get_posts()


@router.get("/{post_id}", response_model=Post)
async def get_post(post_id: int):
    return await controller.get_post(post_id)


@router.get("/{post_id}/comments", response_model=list[Comment])
async def get_post_comments(post_id: int):
    return await controller.get_post_comments(post_id)


@router.post("/", response_model=Post, status_code=status.HTTP_201_CREATED)
async def create_post(post: PostCreate, user: User = Depends(get_current_user)):
    return await controller.create_post(post, user.id)


@router.put("/{post_id}", response_model=Post)
async def update_post(post_id: int, post: PostUpdate):
    return await controller.update_post(post_id, post)


@router.delete("/{post_id}")
async def delete_post(post_id: int):
    return await controller.delete_post(post_id)


@router.get("/{post_id}/votes", response_model=int)
async def get_post_votes(post_id: int):
    return await controller.get_post_votes(post_id)


@router.get("/{post_id}/vote")
async def get_post_vote(post_id: int, user: User = Depends(get_current_user)):
    return await controller.get_post_vote(post_id, user.id)


@router.post("/{post_id}/vote", status_code=status.HTTP_201_CREATED)
async def create_post_vote(
    post_id: int, type: VoteType, user: User = Depends(get_current_user)
):
    return await controller.create_post_vote(post_id, type, user.id)
